bootstrap/deploy_job.yaml

```diff
@@ -0,0 +1,31 @@
+# Deploy kubeflow to namespace "kubeflow"
```

Comment:
Kubectl run would require person to deal with problems that this wont have:
1. GITHUB_TOKEN - first hurdle that's hard to debug
2. ksonnet -> k8s configuration - it's not trivial to run ksonnet from within kubernetes
3. multiple commands, first kubectl run <<image>> bash -> ks generate, ks apply (again, figure out how to run ks in cluster)

vs

kubectl create -f https://raw.githubusercontent.com/kubeflow/kubeflow/master/bootstrap/deploy_job.yaml

Reply:
@nickchase 

I think you misunderstand me. All a K8s job is doing is running a docker container. So everything you are doing with a K8s job you could probably do just by doing kubectl run.

Either way you end up running the same container. So if the container has the Kubeflow registry baked in then they don't need to worry about GITHUB_TOKEN. 

All your deploy_job is doing is running a shell script so the same thing could be done via kubectl run.

The reason I like kubectl run as opposed to 
kubectl create -f ....

is because I think a YAML file is an invitation to end up back where we started. i.e. with a YAML file its tempting to keep adding environment variables to expose more parameters options. Which eventually leads to us picking an appropriate template and packing solution and we're back where we started.

If we do `kubectl run` and/or `docker run` then I think its more likely that we will keep it simple; i.e. restrict the number of parameters to the point its easy to type out the command line.

For anything more complex, I think the solution should be 'here's the ksonnet app created by the bootstraper' now you can customize it a hundred different ways.

So instead of having the bootstrapper expose more parameters to the user (e.g. via a config map). 

I would like to make the bootstrapper
1. better at picking good parameters for the app based on the user's setup
1. making the resulting ksonnet app available to the user (e.g. by pushing to git or a shared filesystem)

Reply:
There is difference in kubectl run vs job - kubectl run is defined by user 100% and job is defined by us - therefore easier.
It won't invalidate kubectl run path - but it will provide very quick, no sweat, way to deploy default environment - perfect for quickstart. kubectl run will require few non-trivial steps.
For kubectl run it's connecting ks with underlying cluster
For docker run it's 1. having docker locally and 2. Exposing kubernetes creds to docker.

Job definition *is not* replacement of your bootstrapper approach, it's alternative or improvement.
If we run bootstraper to do what it's meant to be doing - determining params for deploy, then we can put these params into configmap or something and run this job with it, still less clicks than kubectl run, ks init, ... especially if you don't want to have ksonnet repos baked in.

Reply:
My contention is there is no good default environment. 

The default environment is dependent on the cluster. This the lesson of trying to set the PV for JupyterHub. Not all clusters have default storage class. 

So the point is to run the bootstrapper to determine what the optimal defaults should be.

Please see my other comment. If you want to create a job/statefulset to automatically deploy a cluster I'm fine with it but subject to the following

1. The job should emit the ksonnet app to allow advanced customization
1. The ksonnet app should be accessible after the job finishes
1. The ksonnet app should be created by running the bootstraper.